Add ZenFS, A RocksDB FileSystem for Zoned Block Devices

yhr opened the pull request:
## ZenFS Overview

ZenFS is a simple file system that utilizes RockDB's FileSystem interface to place files into zones on a raw zoned block device. By separating files into zones and utilizing the write life time hints to co-locate data of similar life times, the system write amplification is greatly reduced(compared to conventional block devices) while keeping the ZenFS capacity overhead at a very reasonable level.

ZenFS is designed to work with host-managed zoned spinning disks as well as NVME SSDs with Zoned Namespaces.
For more information on zoned storage devices, go to [ZonedStorage.IO](https://zonedstorage.io/)

Some of the ideas and concepts in ZenFS are based on earlier work done by Abutalib Aghayev and Marc Acosta.

### Performance

By using the zoned storage model to do smart data placement, ZenFS eliminates disk and file system garbage collection which for conventional disks adds 2-4x write amplification. By eliminating garbage collection write throughput and read performance is improved.

Comparing a ZNS drive using ZenFS with using XFS on a comparable conventional drive we see:

* 85% better write performance for 100% overwrite workloads
* Write limited readwhilewriting workloads yielding 43% better read throughput and tail latency reductions by an order of magnitude
* On-par performance for 100% randomread workloads

If sst and write ahead log files are configured to match zone capacity, ZenFS space amplification lands at around 5-15% depending on disk and zone size. Considering that conventional SSDs reserve 7-28% of media to do garbage collection, ZenFS can both reduce write amplification and provide more media storage capacity to the user.

### Current state of implementation

We've spent the last few months evaluating the performance and stabilizing the code - we've run extensive testing on various zone geometries using db_bench and db_stress and done testing with MySQL using MyRocks with ZenFS integrated. The on-disk file format now supports incremental file updates and can now be considered stable.

### Dependencies

ZenFS depends on[ libzbd ](https://github.com/westerndigitalcorporation/libzbd) and Linux kernel 5.4 or later to perform zone management operations.

## Architecture overview
![zenfs stack](https://user-images.githubusercontent.com/447288/84152469-fa3d6300-aa64-11ea-87c4-8a6653bb9d22.png)

ZenFS implements the FileSystem API, and stores all data files on to a raw zoned block device. Log and lock files are stored on the default file system under a configurable directory. Zone management is done through libzbd and ZenFS io is done through normal pread/pwrite calls.

A tool called zenfs is provided in the rocksdb codebase and can be used to manage ZenFS file systems (mkfs, list uuids, list files).

## Example usage

This example issues 100 million random inserts followed by as many overwrites on a 100G memory backed zoned null block device. Target file sizes are set up to align with zone size.


```
make db_bench zenfs

sudo su

DEV=nullb1
ZONE_CAP_SECS=$(blkzone report -c 5 /dev/$FS_PATH | grep -oP '(?<=cap )[0-9xa-f]+' | head -1)
FUZZ=5
ZONE_CAP=$((ZONE_CAP_SECS * 512))
BASE_FZ=$(($ZONE_CAP  * (100 - $FUZZ) / 100))
WB_SIZE=$(($BASE_FZ * 2))

TARGET_FZ_BASE=$WB_SIZE
TARGET_FILE_SIZE_MULTIPLIER=2
MAX_BYTES_FOR_LEVEL_BASE=$((2 * $TARGET_FZ_BASE))

# We need the deadline io scheduler to gurantee write ordering
echo deadline > /sys/class/block/$DEV/queue/scheduler

./zenfs mkfs --zbd=$DEV --aux_path=/tmp/zenfs_$DEV --finish_threshold=$FUZZ --force

./db_bench --fs_uri=zenfs://dev:$DEV --key_size=16 --value_size=800 --target_file_size_base=$TARGET_FZ_BASE --write_buffer_size=$WB_SIZE --max_bytes_for_level_base=$MAX_BYTES_FOR_LEVEL_BASE --max_bytes_for_level_multiplier=4 --use_direct_io_for_flush_and_compaction --max_background_jobs=$(nproc) --num=100000000 --benchmarks=fillrandom,overwrite
```

This graph below shows the capacity usage over time.
As ZenFS does not do any garbage collection the write amplification is 1.

![zenfs_capacity](https://user-images.githubusercontent.com/447288/84157574-2c51c380-aa6b-11ea-89e2-def4a4d658af.png)

## File system mounting

A ZenFS file system may be mounted by pointing to a zoned block device (zenfs://dev:<zoned block device>) or by specifying the file system uuid by providing uris (zenfs://uuid:<uuid>). The URIs are provided when loading the file system using FileSystem::Load or as the --fs_uri parameters to db_bench and db_stress. UUIDs can be listed by zenfs tool using the ls-uuid command.

## File system implementation

Files are mapped into into a set of extents:

* Extents are block-aligned, continious regions on the block device
* Extents do not span across zones
* A zone may contain more than one extent
* Extents from different files may share zones

### Reclaim 

ZenFS is exceptionally lazy at current state of implementation and does not do any garbage collection whatsoever. As files gets deleted, the used capacity zone counters drops and when it reaches zero, a zone can be reset and reused.

###  Metadata 

Metadata is stored in a rolling log in the first zones of the block device.

Each valid meta data zone contains:

* A superblock with the current sequence number and global file system metadata
* At least one snapshot of all files in the file system
* Incremental file system updates (new files, new extents, deletes, renames etc)

yhr commented:
Force-pushed a bunch of small fixups to zenfs-pullreq

yhr commented:
@ajkr : Do you guys want some additional documentation or testing to be done before digging into the code? We'd like to see the zoned support eventually making it into in myrocks, so it would be good to know if there are any blockers. CI seems happy with the changes (the failed travis check is as far as I can see unrelated). Cheers!

pdillinger commented:
This looks like great work. However, until it's considered widely applicable, the cost of having the extra code around in the main repository (ongoing maintenance of RocksDB) likely exceeds the general benefit. And since this is implemented as a very clean extension (👏 ), there don't appear to be RocksDB core features that are missing to make this extension workable and performant. The Western Digital fork seems like the natural place for this until there's a case for broad applicability.

Or is this applicable with generic HDD or SSD? (It didn't appear so to me.)

yhr commented:
@pdillinger : thanks for your feedback. I would argue that zoned support IS widely applicable as:

* In addition to supporting NVMe ZNS SSDs, the backend also supports host-managed SMR HDDs. SMR HDDs are deployed in production in many large data centers and have been around for many years.
* There is an extraordinary amount of work in academia that shows the benefits of RocksDB when integrating SMR HDDs, ZNS SSDs, Open-Channel SSDs, and similar storage architectures. However, everyone continues to reinvent the work, as there isn't a standard implementation that can be continuously improved. This work finally takes a step forward and unifies all of the research in the industry and academia, and now enables production systems to get the benefits.
* Both the Linux kernel and SPDK has upstream support for zoned block devices. Additionally, there is quite a bit of tools and libraries that support these devices as well.
* The integration plugs easily into downstream users (e.g. MyRocks) and shows direct performance benefits. Comparing MyRocks/MySQL (on top of XFS) with ZenFS, it can execute 41% more transactions in the oltp_writeonly.lua workload and 61% more transactions in the oltp_readwrite.lua workload.

If there is a concern that this would be a "drop'n'run" PR - I'll certainly help out doing reviews, CI-integration and will continue to be an active maintainer on this piece of work. I have a long history in the kernel community to participate, and will also continue work on this backend.

We're supporting downstream users with the adoption, but they do rely on upstream support. RocksDB releases take time to trickle down, so it is key to get support in early.

To make full use of zoned storage devices, applications need adoption to make use of smart data placement. Zoned support in RocksDB is a great piece in this puzzle.

yhr commented:
@mrambacher @anand1976 : Would you have time to look this over? Your feedback would be very welcome - both on the architecture and the implementation.I've tried to minimize the impact and the maintenance effort, so and I really hope that adding support for a new class of storage devices in rocksdb makes up for the added lines of code. Cheers!

mrambacher commented:
@yhr I will try to look it over this week.  I really like the idea!

ajkr reviewed (commented):
Thanks very much for the PR. I left some minor comments/questions on the details but also have some high-level questions/comments.

Have you tried any space efficiency optimizations? I think it is an exciting opportunity for zoned SSDs to save the space typically reserved for internal GC. For example, I wonder if we could save most of that space by aligning file size to zone size. We can set `target_file_size_base` to the zone size and `target_file_size_multiplier == 1` to achieve it for files in L1+. We can't really do the same for L0 files since those file sizes are determined by `write_buffer_size` (i.e., memtable size), which is a bound on uncompressed data size, while L0 files will be compressed when written. Although, L0 files are typically deleted in the same order they are written (with some exceptions), so maybe it is fine to just write them out consecutively.

The above is kind of tricky as it imposes a restriction for file size to match zone boundary. Plus, the `target_file_size_base` is after all just a target for the file size, and it may be undershot/overshot in practice, sometimes severely. It might still be worth trying though. I am developing another idea to make L1+ levels more like L0 in the sense that file creation and deletion occur in the same order (LevelDB actually picked files for compaction in a round robin style within a level and we got rid of that for write-amp reasons, but IMO we can save the write-amp in a different way while preserving round robin ordering). Then, we can dedicate zones to a level and write out files consecutively within that level's zones (this requires us to tell `FileSystem` the file's level). Since deletion happens in the same order as creation for a given level, deleted data will be consecutive within a given level's zones. I think that'd allow us to bound the number of zones lost to overhead per level without depending on file size.

We should discuss internally more about where the code should live. There is precedent for supporting specific platforms in the main repo, like libhdfs and librados. I also wonder whether plugins like those or this one could live in a separate repo. Besides dependencies on non-public APIs (it looks like this one uses `PutFixed*()`, `GetFixed*()`, and crc32c calculation, which should be replaceable), I am not sure if there are other reasons a fork would be needed.

Also if we do include in RocksDB repo, we should probably rename the files along the lines of "env/fs_libzbd.h". We get a lot of complaints about code size so try to name such that people can quickly determine whether code is relevant to them without needing to read any of it.

yhr commented on env/fs_zenfs.h:
```diff
@@ -0,0 +1,413 @@
+    nr_zones_ = zbd->GetNrZones();
+
+    assert(aux_fs_path.length() < sizeof(aux_fs_path_));
+    strcpy(aux_fs_path_, aux_fs_path.c_str());
```
Yep, and add a check for the length in mkfs

yhr commented on env/fs_zenfs.h:
```diff
@@ -0,0 +1,413 @@
+  IOStatus GetAbsolutePath(const std::string& db_path, const IOOptions& options,
+                           std::string* output_path,
+                           IODebugContext* dbg) override {
+    return target()->GetAbsolutePath(db_path, options, output_path, dbg);
```
Yes, i think that would make sense

yhr commented on env/fs_zenfs.h:
```diff
@@ -0,0 +1,413 @@
+                                      const FileOptions& options,
+                                      std::unique_ptr<FSWritableFile>* result,
+                                      IODebugContext* dbg) {
+    return target()->NewWritableFile(fname, options, result, dbg);
```
Yes, and we should probably check if the file exists in zenfs first

yhr commented on env/fs_zenfs.cc:
```diff
@@ -0,0 +1,973 @@
+}
+
+void ZenFS::ClearFiles() {
+  std::map<std::string, ZoneFile*>::iterator it;
```
This is only done during recovery, but might as well grab the lock to avoid any future issues

yhr commented on env/fs_zenfs.cc:
```diff
@@ -0,0 +1,973 @@
+  files_.clear();
+}
+
+IOStatus ZenFS::WriteSnapshot(ZenMetaLog* meta_log) {
```
Yep, a rename and some documentation would not hurt

yhr commented on env/fs_zenfs.cc:
```diff
@@ -0,0 +1,973 @@
+IOStatus ZenFS::WriteEndRecord(ZenMetaLog* meta_log) {
+  std::string endRecord;
+
+  PutFixed32(&endRecord, kCompleteFilesSnapshot);
```
Indeed, nice catch, thanks.

yhr commented on env/fs_zenfs.cc:
```diff
@@ -0,0 +1,973 @@
+  s = meta_log_->AddRecord(record);
+  if (s == IOStatus::NoSpace()) {
+    Info(logger_, "Current meta zone full, rolling to next meta zone");
+    s = RollMetaZone();
```
Yep, should be documented

yhr commented on env/fs_zenfs.cc:
```diff
@@ -0,0 +1,973 @@
+
+    zoneFile = files_[fname];
+    EncodeFileDeletionTo(zoneFile, &record);
+    s = PersistRecord(record);
```
Yeah, i'm going to look this over. We're going to loose the delete on a roll as it is now. Either we write records after the roll, or delete the file from files_ and re-insert it if the roll fails. Thanks.

yhr commented on env/fs_zenfs.cc:
```diff
@@ -0,0 +1,973 @@
+      valid_logs[i].reset();
+    }
+  }
+  s = RollMetaZone();
```
Yes, I'll do a pass over this and add documentation. Thanks for digging into it.

yhr commented:
@ajkr Thanks a lot for the review! I'll post an push an update next week.

I really like the idea of the round-robin compaction. That would be the optimal way to utilize the space while keeping write amp at 1. Right now i'm doing fuzzy file size alignment to deal with files not matching zone capacities (setting target file sizes to 90% of zone capacity and finishing unused zones with less than 10% left). This works - for 1TB drives the space amp can be reduced to as little as 5-6% - but is not as robust as I'd like.

yhr commented:
Pushed an update addressing the comments from @ajkr . Did not rename the files yet - *_libzbd.* would bit misleading as libzbd is just a thin wrapper of the kernel interfaces  - maybe *_zoned.*?

As the file system interface API is still being worked on, i think it would make a whole lot sense to include the code rocksdb repo to avoid breakage. I'm also concerned about downstream users having to deal with different repos. Maybe a new directory structure(adding e.g fs/zenfs/) and some added documentation would make things clear to anyone going through the code?

ajkr commented:
> i think it would make a whole lot sense to include the code rocksdb repo to avoid breakage.

The `FileSystem` interface is considered part of the public API so changes have to be compatible (except for major version upgrades but we don't do much of those), otherwise it's a bug on our side.

> I'm also concerned about downstream users having to deal with different repos.

It is a good point. I think @mrambacher's work will make it easier to use custom plugins. Mark, do you have any opinion on the separate repo approach?

mrambacher reviewed (commented):
I have not been through all of the code yet but here are the comments I have so far.  Some documentation as to what some of the functions do would be helpful (more comments please!).

As far as where the code lives, I have started an effort to make "customizable extensions" easier to add to RocksDB.  My goal is to make extensions such as this one pluggable via dynamic libraries and options files/configurations into existing RocksDB projects.  I will push that conversation along to see how people would like to see these extensions evolve.

yhr commented on env/fs_posix.cc:
```diff
@@ -1055,8 +1056,42 @@ static FactoryFunc<FileSystem> posix_filesystem_reg =
+          Status s;
+
+          devID.replace(0, strlen("zenfs://"), "");
+          if (devID.rfind("dev:") == 0) {
+            devID.replace(0, strlen("dev:"), "");
+            s = NewZenFS(&fs, devID);
```
It is not needed, as the dev: mode it directly names a zoned block device.  I think it is useful to keep it like this, in case of ListZenFileSystems() failing for some reason.

yhr commented on env/fs_zenfs.h:
```diff
@@ -0,0 +1,413 @@
+  Status MkFS(std::string aux_fs_path, uint32_t finish_threshold);
+
+  const char* Name() const override {
+    return "ZenFS - The Zoned-enabled File System";
```
sure, that's more concise

yhr commented on env/fs_zenfs.h:
```diff
@@ -0,0 +1,410 @@
+};
+#endif  // !defined(ROCKSDB_LITE) && defined(OS_LINUX) && defined(LIBZBD)
+
+Status NewZenFS(FileSystem** fs, const std::string& bdevname);
```
The registred factory function takes a std::unique_ptr<FileSystem>*, I presume that this is what you want reflected here?

mrambacher replied:
No, I actually meant a std::shared_ptr<FileSystem>,  I believe that is what is stored inside the Env object and is what you should return here.  The factory guard is more of a "transient" entity.

yhr replied:
But the FileSystem pointer is turned into a shared one in NewSharedObject when loading? Am I confused?

yhr commented on env/io_zenfs.cc:
```diff
@@ -0,0 +1,615 @@
+//  COPYING file in the root directory) and Apache 2.0 License
+//  (found in the LICENSE.Apache file in the root directory).
+
+#if !defined(ROCKSDB_LITE) && !defined(OS_WIN) && defined(LIBZBD)
```
No reason, the guard should match fs_zenfs.h

yhr commented on env/fs_zenfs.h:
```diff
@@ -0,0 +1,410 @@
+
+  /* Create a superblock for a filesystem covering the entire zoned block device
+   */
+  Superblock(ZonedBlockDevice* zbd, std::string aux_fs_path = "",
```
fixed

yhr commented on env/fs_zenfs.h:
```diff
@@ -0,0 +1,410 @@
+   */
+  Superblock(ZonedBlockDevice* zbd, std::string aux_fs_path = "",
+             uint32_t finish_threshold = 0) {
+    std::string uuid = Env::Default()->GenerateUniqueId();
```
I'm referring to the default env (not filesystem)  so this  should be fine?

yhr commented on env/fs_zenfs.h:
```diff
@@ -0,0 +1,410 @@
+
+  /* Create a superblock for a filesystem covering the entire zoned block device
+   */
+  Superblock(ZonedBlockDevice* zbd, std::string aux_fs_path = "",
```
Might as well move this over to fs_zenfs.cc

yhr commented on env/fs_zenfs.h:
```diff
@@ -0,0 +1,410 @@
+#if !defined(ROCKSDB_LITE) && defined(OS_LINUX) && defined(LIBZBD)
+
+class Superblock {
+  uint32_t magic_ = 0;
```
yep.

yhr commented on env/fs_zenfs.h:
```diff
@@ -0,0 +1,410 @@
+    return Status::OK();
+  }
+
+  uint32_t GetSeq() { return sequence_; }
+  std::string GetAuxFsPath() { return std::string(aux_fs_path_); }
+  uint32_t GetFinishTreshold() { return finish_treshold_; }
+  std::string GetUUID() { return std::string(uuid_); }
```
That makes sense

yhr commented on env/fs_zenfs.h:
```diff
@@ -0,0 +1,410 @@
+
+  uint32_t GetSeq() { return sequence_; }
+  std::string GetAuxFsPath() { return std::string(aux_fs_path_); }
+  uint32_t GetFinishTreshold() { return finish_treshold_; }
```
yes!

yhr commented on env/fs_zenfs.h:
```diff
@@ -0,0 +1,410 @@
+  IOStatus AddRecord(const Slice& slice);
+  IOStatus ReadRecord(Slice* record, std::string* scratch);
+
+  Zone* GetZone() { return zone_; };
```
fixed

yhr commented on env/fs_zenfs.h:
```diff
@@ -0,0 +1,410 @@
+};
+
+class ZenMetaLog {
+  uint64_t read_pos_;
```
fixed

yhr commented on env/zbd_zenfs.h:
```diff
@@ -0,0 +1,120 @@
+  unsigned int max_nr_open_io_zones_;
+
+ public:
+  explicit ZonedBlockDevice(std::string bdevname,
+                            std::shared_ptr<Logger> logger);
```
fixed

yhr commented on env/zbd_zenfs.cc:
```diff
@@ -0,0 +1,510 @@
+//  COPYING file in the root directory) and Apache 2.0 License
+//  (found in the LICENSE.Apache file in the root directory).
+
+#if !defined(ROCKSDB_LITE) && !defined(OS_WIN) && defined(LIBZBD)
```
fixed, i checked the rest of the files as well

yhr commented on env/zbd_zenfs.cc:
```diff
@@ -0,0 +1,510 @@
+
+#if !defined(ROCKSDB_LITE) && !defined(OS_WIN) && defined(LIBZBD)
+
+#include "zbd_zenfs.h"
```
fixed, also made a pass over the rest of the files fixing up the include paths

yhr commented on env/zbd_zenfs.cc:
```diff
@@ -0,0 +1,510 @@
+#include "io_zenfs.h"
+#include "rocksdb/env.h"
+
+#define KB (1024)
+#define MB (1024 * KB)
```
These are not defined elsewhere. In stead  (1024 * 1024 ) for megabyte constants, so I dropped these.

yhr commented on env/zbd_zenfs.cc:
```diff
@@ -0,0 +1,510 @@
+
+  if (info.nr_zones < ZENFS_MIN_ZONES) {
+    return IOStatus::NotSupported(
+        "To few zones on zoned block device (32 required)");
```
fixed

yhr commented on env/zbd_zenfs.cc:
```diff
@@ -0,0 +1,510 @@
+  return nullptr;
+}
+
+ZonedBlockDevice::ZonedBlockDevice(std::string bdevname,
+                                   std::shared_ptr<Logger> logger)
+    : filename_("/dev/" + bdevname), logger_(logger) {
```
fixed

yhr commented on env/fs_zenfs.cc:
```diff
@@ -0,0 +1,1013 @@
+  ZonedBlockDevice* zbd = new ZonedBlockDevice(bdevname, logger);
+  IOStatus zbd_status = zbd->Open();
+  if (!zbd_status.ok()) {
+    Error(logger, "Failed to open zoned block device: %s",
+          zbd_status.ToString().c_str());
+    return Status::IOError(zbd_status.ToString());
+  }
```
indeed, fixed

yhr commented on env/fs_zenfs.cc:
```diff
@@ -0,0 +1,1013 @@
+          }
+        }
+
+        delete zbd;
```
correct, fixed

yhr commented on env/fs_zenfs.cc:
```diff
@@ -0,0 +1,1013 @@
+}
+#endif
+
+Status NewZenFS(FileSystem** fs, const std::string& bdevname) {
```
That makes sense

yhr commented on env/fs_zenfs.cc:
```diff
@@ -0,0 +1,1013 @@
+  while (NULL != (entry = readdir(dir))) {
+    if (entry->d_type == DT_LNK) {
+      std::string zbdName = std::string(entry->d_name);
+      ZonedBlockDevice* zbd = new ZonedBlockDevice(zbdName, nullptr);
+      IOStatus zbd_status = zbd->Open(true);
```
no, better stash it on the stack and avoid the cleanup work

yhr commented on env/fs_zenfs.cc:
```diff
@@ -0,0 +1,1013 @@
+
+        for (const auto z : metazones) {
+          Superblock super_block;
+          std::unique_ptr<ZenMetaLog> log;
+          log.reset(new ZenMetaLog(zbd, z));
```
no, creating it on the stack is better

yhr commented on env/fs_zenfs.cc:
```diff
@@ -0,0 +1,1013 @@
+
+std::map<std::string, std::string> ListZenFileSystems() {
+  std::map<std::string, std::string> zenFileSystems;
+  DIR* dir = opendir("/sys/class/block");
```
yes it should be closed, fixed

yhr commented on env/fs_zenfs.h:
```diff
@@ -0,0 +1,413 @@
+    return "ZenFS - The Zoned-enabled File System";
+  }
+
+  virtual IOStatus NewSequentialFile(const std::string& fname,
+                                     const FileOptions& file_opts,
```
fixed

yhr commented:
@mrambacher  : Thanks for the review! I just pushed updates based on your comments (except for a couple of things that are still marked as unresolved, see above)

yhr commented:
@zhichao-cao : Have you had a chance to look at the code?

zhichao-cao commented:
> @zhichao-cao : Have you had a chance to look at the code?

Yeah, one initial question. Did you run the experiment on a real device or on QEMU?

yhr commented:
> Yeah, one initial question. Did you run the experiment on a real device or on QEMU?

To get the comparison numbers, we ran the same workloads on both a zns drive and a conventional drive (with the same hw platform/media setup, but different FW)

yhr commented:
ZenFS is now available as a build-time plugin from https://github.com/westerndigitalcorporation/zenfs where it will be maintained going forward.

yhr commented:
ZenFS is now available as a build-time plugin from https://github.com/westerndigitalcorporation/zenfs where it will be maintained going forward. See the README in the repository for more information. Thanks for all the great feedback everyone!
